### PR TITLE
Export WitnessCollector in the RPC interface

### PR DESCRIPTION
**Pull request by stdevMac**
## Changes:
- Add WitnessRpcModule and tests for it

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

## Further comments (optional)

Export WitnessCollector in the RPC interface

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/Modules/ModuleType.cs [hunk 1]
```diff
@@ -85,7 +87,8 @@ public static class ModuleType
             Proof, 
             Net,
             Parity,
-            Health
+            Health,
+            Witness
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/ModuleType.cs:91]: Not sure if we want to make it default module, doesn't really sound crucial

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/IWitnessRpcModule.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        [JsonRpcMethod(Description = "Return last N witness blocks",
+            EdgeCaseHint = "",
+            ResponseDescription = "Last n block witnesses",
+            ExampleResponse =
+                "[\"0xe686ead4169241f99b318240a0557ab5dd4d99444367ca4a28e60cda5717ef2c\"]",
+            IsImplemented = true)]
+        Task<ResultWrapper<string>> get_witnesses(
+            [JsonRpcParameter(Description = "Number of witness blocks to return", ExampleValue = "1")]
+            string n);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/IWitnessRpcModule.cs:33]: I thinkBlockParameter will be better and more consistent with other RPC
also should be at least an int, not a string

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+using Nethermind.Core.Crypto;
+using Nethermind.State;
+
+namespace Nethermind.JsonRpc.Modules.Witness
+{
+    public class WitnessRpcModule : IWitnessRpcModule
+    {
+        private readonly IWitnessCollector? _wrapper;
+
+        public WitnessRpcModule(IWitnessCollector? wrapper)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:31]: use by IWitnessRepository

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            _wrapper = wrapper;
+        }
+
+        public async Task<ResultWrapper<string>> get_witnesses(string n)
+        {
+            if (!int.TryParse(n, out int numberOfBlocks) || _wrapper is null)
+                return ResultWrapper<string>.Fail("Can convert n (represent the number of witness to return) to int");
+            IEnumerable<Keccak> collected = _wrapper.Collected.Skip(Math.Max(0, _wrapper.Collected.Count - numberOfBlocks));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:40]: You should use IWitnessRepository.Load for the block not current collected witness for current block

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Nethermind.Core.Crypto;
+using Nethermind.State;
+
+namespace Nethermind.JsonRpc.Modules.Witness
+{
+    public class WitnessRpcModule : IWitnessRpcModule
+    {
+        private readonly IWitnessCollector? _wrapper;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:29]: wrapper?

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/IWitnessRpcModule.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+using Nethermind.Blockchain.Find;
+using Nethermind.Core.Crypto;
+
+namespace Nethermind.JsonRpc.Modules.Witness
+{
+    [RpcModule(ModuleType.Witness)]
+    public interface IWitnessRpcModule : IRpcModule
+    {
+        [JsonRpcMethod(Description = "Return witness of Block provided",
+            ResponseDescription = "Keccak representation of Witness",
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/IWitnessRpcModule.cs:28]: ResponseDescription = "Table of hashes of state nodes that were read during block processing"

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/IWitnessRpcModule.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            ExampleResponse =
+                "[\"0xe686ead4169241f99b318240a0557ab5dd4d99444367ca4a28e60cda5717ef2c\"]",
+            IsImplemented = true)]
+        Task<ResultWrapper<Keccak[]>> get_witnesses([JsonRpcParameter(Description = "Block to get witness",
+                ExampleValue = "[\"8934677\"]")]
+            string blockHash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/IWitnessRpcModule.cs:34]: This should be BlockParameter like here in eth_getCode: https://github.com/NethermindEth/nethermind/blob/3161505177f0104236e4b2525a0fd029a83498b4/src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs#L273

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+        public WitnessRpcModule(IWitnessRepository witnessRepository, IBlockFinder finder)
+        {
+            _witnessRepository = witnessRepository;
+            _blockFinder = finder;
+        }
+
+        public async Task<ResultWrapper<Keccak[]>> get_witnesses(string blockHash)
+        {
+            var blockParameter = new BlockParameter(new Keccak(blockHash));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:39]: this is wrong, you don't expect hash only

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            Block block = _blockFinder.FindBlock(blockParameter);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:40]: Header will suffice (optimization)
You also need to handle situation where its not available and returns a null.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            Keccak[] result = _witnessRepository.Load(block.Hash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:41]: You need to handle situation where _witnessRepository returns a null (witness is not available)

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.JsonRpc.Test/Modules/WitnessModuleTests.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            _blockFinder.FindBlock((BlockParameter)null).ReturnsForAnyArgs(_block);
+
+            _witnessRepository = new WitnessCollector(new MemDb(), LimboLogs.Instance);
+            
+            _witnessRepository.Add(_block.Hash);
+            _witnessRepository.Persist(_block.Hash);
+            _witnessRpcModule = new WitnessRpcModule(_witnessRepository, _blockFinder);
+        }
+
+        [Test]
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc.Test/Modules/WitnessModuleTests.cs:61]: Add test for missing data

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs (hunk 9 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:40]: @stdevMac -> _blockFinder has a FindHeader method which returns a header only, without a block body (without transactions) which is much lighter on the DB

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+        public WitnessRpcModule(IWitnessRepository witnessRepository, IBlockFinder finder)
+        {
+            _witnessRepository = witnessRepository;
+            _blockFinder = finder;
+        }
+
+        public async Task<ResultWrapper<Keccak[]>> get_witnesses(BlockParameter blockParameter)
+        {
+            BlockHeader blockHeader = _blockFinder.FindHeader(blockParameter);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:40]: Use SearchForHeader extension method for better error handling

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            if (blockHeader is null)
+                return ResultWrapper<Keccak[]>.Fail("Block not found");
+            Keccak hash = blockHeader.CalculateHash();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:43]: use blockHeader.Hash, don't recalculate it

File: src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+            Keccak[] result = _witnessRepository.Load(hash);
+            return result is null ? ResultWrapper<Keccak[]>.Fail("Witness not found") : ResultWrapper<Keccak[]>.Success(result);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Witness/WitnessRpcModule.cs:45]: add error code ErrorCodes.ResourceUnavailable

File: src/Nethermind/Nethermind.JsonRpc.Test/Modules/WitnessModuleRpcTests.cs [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+                RpcTest.TestSerializedRequest<IWitnessRpcModule>(_witnessRpcModule, "get_witnesses", _block.CalculateHash().ToString());
+            serialized.Should().Be(GetOneWitnessHashResponse);
+        }
+
+        [Test]
+        public void BlockNotFound()
+        {
+            _blockFinder = Substitute.For<IBlockTree>();
+            _witnessRepository = new WitnessCollector(new MemDb(), LimboLogs.Instance);
+            _witnessRpcModule = new WitnessRpcModule(_witnessRepository, _blockFinder);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc.Test/Modules/WitnessModuleRpcTests.cs:85]: Setup could be extracted

**Review by LukaszRozmej — APPROVED** [review]
Generally ok, but please improve error handling

**Pull request merged**